File: yt/yt/flow/examples/kotlin/external_state_join/external_state_join/src/main/kotlin/tech/ytsaurus/flow/examples/externalstatejoin/LookupJoin.kt
package tech.ytsaurus.flow.examples.externalstatejoin

import tech.ytsaurus.flow.computation.OutputCollector
import tech.ytsaurus.flow.context.RuntimeContext
import tech.ytsaurus.flow.function.RowFunction
import tech.ytsaurus.flow.row.ExtendedMessage
import tech.ytsaurus.flow.state.JoinedExternalStateDescriptor
import tech.ytsaurus.flow.state.StateDescriptors

/**
 * Joins each event against a pre-built dynamic reference table reached through
 * the `/reference` external state. The external state joiner resolves the Cypress
 * symlink behind that path per lookup, so repointing the symlink swaps the
 * reference dataset for keys looked up afterwards without a pipeline restart.
 */
class LookupJoin : RowFunction {
    companion object {
        private val REFERENCE_STATE: JoinedExternalStateDescriptor =
            StateDescriptors.externalReadOnly("/reference")
    }

    // [BEGIN on_message]
    override fun onMessage(message: ExtendedMessage, output: OutputCollector, ctx: RuntimeContext) {
        val stateAccessor = ctx.getState(REFERENCE_STATE, message)
        val name = stateAccessor.orDefault.get("name", String::class.java) ?: return

        output.addMessage(
            ctx.createMessageBuilder("enriched")
                .set("key", message.get("key", Long::class.java))
                .set("name", name)
                .finish()
        )
    }
    // [END on_message]
}

File: yt/yt/flow/examples/go/external_state_join/lookup_join.go
package main

import (
	"context"
	"errors"

	"go.ytsaurus.tech/yt/go/flow"
)

const (
	referenceStateName = "/reference"
	enrichedStreamID   = "enriched"
)

// [BEGIN lookup_join]

type referenceState struct {
	Name *string `yson:"name"`
}

type lookupJoin struct{}

var _ flow.RowFunction = (*lookupJoin)(nil)

func (*lookupJoin) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	var event eventMessage
	if err := msg.ConvertTo(&event); err != nil {
		return err
	}

	name, ok, err := referenceName(rt, msg)
	if err != nil || !ok {
		return err
	}

	enriched := flow.NewYSONMessage[enrichedMessage](enrichedStreamID)
	enriched.Key = event.Key
	enriched.Name = name
	encoded, err := flow.ConvertFrom(rt, enriched)
	if err != nil {
		return err
	}
	out.AddMessage(encoded)
	return nil
}

func referenceName(rt flow.Runtime, msg flow.ExtendedMessage) (string, bool, error) {
	reference, err := flow.OpenJoinedExternalState(rt, referenceStateName, msg)
	if errors.Is(err, flow.ErrStateNotRead) {
		return "", false, nil
	}
	if err != nil {
		return "", false, err
	}

	var state referenceState
	exists, err := reference.ConvertTo(&state)
	if err != nil || !exists || state.Name == nil {
		return "", false, err
	}
	return *state.Name, true, nil
}

// [END lookup_join]
